Update backend [part 4]

r0mainK opened the pull request:
Take out unused methods in backend relating to the index, add `initialize_backend` method, create helper function in gcs_backend to avoid redundancy

vmarkovtsev commented:
@r0mainK There are merge conflicts

r0mainK commented:
@vmarkovtsev solved them, ready for review when you get the time

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -24,19 +23,10 @@ def publish_model(args: argparse.Namespace, backend: StorageBackend, log: loggin
-    except Exception as e:
-        log.critical("Failed to load the model: %s: %s" % (type(e).__name__, e))
-        return 1
+    model = GenericModel(source=path, dummy=True)
```
Why did you remove the error handlers? It is a top level function

r0mainK replied:
here was my mistake, reverting

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -24,19 +23,10 @@ def publish_model(args: argparse.Namespace, backend: StorageBackend, log: loggin
-        model_url = backend.upload_model(path, base_meta, args.force)
-    except ModelExistsError:
-        return 1
+    model_url = backend.upload_model(path, base_meta, args.force)
```
Same here

r0mainK replied:
here I don't understand why, since the error is a custom one that clearly explains whats happening - and logs in `upload_model` also provide information. Unless you want to sytematically wrap errors this way to return 1 ? In which case we should wrap the other top level functions too, no ?

r0mainK replied:
anyway reverting

vmarkovtsev replied:
Yep we should wrap everything, that was a dirty cutting off the edges

r0mainK commented on modelforge/gcs_backend.py:
```diff
@@ -75,31 +75,41 @@ def bucket_name(self):
+        client = self.create_client()
         return client.lookup_bucket(self.bucket_name)
 
+    def initialize_backend(self, force):
```
done

r0mainK commented on modelforge/gcs_backend.py:
```diff
@@ -75,31 +75,41 @@ def bucket_name(self):
         if blob.exists() and not force:
             self._log.error("Model %s already exists, aborted.", meta["uuid"])
-            return 1
+            raise ExistingModelError
```
done

r0mainK commented on modelforge/gcs_backend.py:
```diff
@@ -124,23 +134,14 @@ def request_with_progress(method, url, data=None, headers=None, **kwargs):
-        if isinstance(where, str):
-            os.makedirs(os.path.dirname(where), exist_ok=True)
-            f = open(where, "wb")
+    def fetch_model(self, source: str, file: str, chunk_size: int =DEFAULT_CHUNK_SIZE) -> None:
```
my bad, changing

r0mainK replied:
also added typing to the rest of the file

r0mainK commented on modelforge/gcs_backend.py:
```diff
@@ -124,23 +134,14 @@ def request_with_progress(method, url, data=None, headers=None, **kwargs):
-            f = open(where, "wb")
+    def fetch_model(self, source: str, file: str, chunk_size: int =DEFAULT_CHUNK_SIZE) -> None:
+        self._log.info("Fetching %s...", source)
+        r = requests.get(source, stream=True)
```
done

r0mainK commented:
Added catches about everywhere, and moved the message to the logger in most cases. I put two TODO for when dulwich implements a porcelain error. Gonna add some test now, to get coverage a bit higher. But yeah ready for more review or merge after that. 
I'm also gonna reorder the methods and functions so they are more coherent.

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -6,12 +6,56 @@
+        return 1
+
+    log.info("Deleting content of index ...")
+    backend.index.initialize_index()
```
Future commit: rename `initialize_index` to `reset`

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -6,12 +6,56 @@
+    log.info("Deleting content of index ...")
+    backend.index.initialize_index()
+    try:
+        backend.index.upload_index("initilialize", {})
```
Future commit: typo

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -6,12 +6,56 @@
+    except ExistingBackendError:
+        return 1
+
+    log.info("Deleting content of index ...")
```
Future commit: "Resetting the index"

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -111,11 +135,16 @@ def delete_model(args: argparse.Namespace, backend: StorageBackend, log: logging
+        template_readme = backend.index.load_template(args.template_readme)
+        backend.index.update_readme(template_readme)
+    except ValueError:
+        return 1
```
We are catching much code here, so better to split and log what has failed exactly.

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -111,11 +135,16 @@ def delete_model(args: argparse.Namespace, backend: StorageBackend, log: logging
-    backend.index.update_readme(template_readme)
-    backend.index.upload_index("delete", meta)
+    try:
+        backend.index.upload_index("delete", meta)
```
Future commit: rename to `upload`
